language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -172,14 +172,14 @@ kindOf (TyApp x ty ty') = do
 intApp :: Type a () -> Natural -> Type a ()
 intApp ty n = TyApp () ty (TyInt () n)
```

Comment:
Since this is used only to apply a type to a `TyInt` (the name is misleading as hell) which almost always means we're applying a `TyBuiltin` (and that's the only usage of the non-exported `intApp`), it probably should be

```haskell
intApp :: TypeBuiltin a -> Size -> NormalizedType name a
intApp ty n = NormalizedType $ TyApp () (TyBuiltin () ty) (TyInt () n)
```

(haven't actually tried to type check the code)

Reply:
I deliberately didn't do this, because it could produce a non-normalized type if applied to a lambda. We don't do that, but I thought it was better to make the lack of safety obvious.

Reply:
> if applied to a lambda

But it's always applied to a `TypeBuiltin`.

Reply:
> We don't do that, but I thought it was better to make the lack of safety obvious.

Not a big deal, though.

Reply:
Sorry, I do not understand. What is unsafe about `intApp`? I mean it literally receives a `TypeBuiltin`.

Reply:
It takes a `Type`. I'm just being paranoid about future users who might pass in a lambda!

Reply:
My version takes a `TypeBuiltin`. That's why I suggested it like that in the first place.

Reply:
Oh, my fault, I failed to read your code. That does indeed seem better, sorry!

Reply:
Hopefully you're going to replace all the builtin stuff with references to the things in `Constant` anyway :)

Reply:
Oh I didn't realize it's me who are going to replace them :)
I'll ask Vanessa in her PR. I think if all three of us are going to modify the same piece of code, it will be overpopulated.

This is indeed a pointless improvement then.